libkbfs/file_data.go

```diff
@@ -593,52 +878,65 @@ func (fd *fileData) write(ctx context.Context, data []byte, off int64,
 		nCopied += fd.bsplit.CopyUntilSplit(
 			block, nextBlockOff < 0, data[nCopied:max], off+nCopied-startOff)
 
-		// TODO: support multiple levels of indirection.  Right now the
-		// code only does one but it should be straightforward to
-		// generalize, just annoying
-
-		// if we need another block but there are no more, then make one
+		// If we need another block but there are no more, then make one.
 		switchToIndirect := false
-		if nCopied < n && nextBlockOff < 0 {
-			// If the block doesn't already have a parent block, make one.
-			if ptr == fd.rootBlockPointer() {
-				topBlock, err = fd.createIndirectBlock(
-					df, DefaultNewBlockDataVersion(false))
+		if nCopied < n {
+			needExtendFile := nextBlockOff < 0
+			needFillHole := off+nCopied < nextBlockOff
+			newBlockOff := startOff + int64(len(block.Contents))
+			if newBlockOff < off+nCopied {
```

Comment:
Consider a file with data `data` with leaf blocks split up like so `data[0:2], data[2:4], data[4], hole(10)`.  Then a write comes in for two bytes at index 8.  (This is `TestFileDataWriteEndOfHole` in `file_data_test.go`.)

The first thing it tries to do is write to the block starting at `startOff=4`.  But since the write doesn't start until offset 8, `CopyUntilSplit` just fills out that block to capacity, but returns `0` for `nCopied`.  `off=8`, so `6 < 8+0` still, and so we need to make the new right block start at `8`.

Perhaps this only ever happens if `nCopied` is 0, so that part of the formula is useless?  Would deleting that, plus adding a better comment, make it more clear?